### ISSUE TITLE
Discrete Material System has design issues.

### ISSUE THREAD

**Issue by dschwen** [idaholab/moose#21847]
## Reason
Example: You are outputting material properties from a discrete material that is called 

a) through a custom API (i.e. MOOSE has no idea how to update its properties), or 
b) meant to be called by two different compute-true materials. 

a) if we do not compute the calling material the discrete material will not be updated, and there is no way of doing that because the calling material is using a custom API instead of `computePropertiesAtQp`. 
b) since two materials use the same discrete material, we do not know which is the state of the properties that need to be outputted.

## Design
Requires discussion with @idaholab/moose-ccb 

## Impact
Enable #21804 and improve discrete material system.

**Comment by GiudGiud** [comment]
If MOOSE does not know how to update the properties you should write your own auxkernel to output the property, that has the custom call that you need

**Comment by hugary1995** [comment]
For a), even if the property is computed by `computePropertiesAtQp`, we may still not be able to output it, as that property could be a function of other properties from compute-true materials. In other words, we rely on `computePropertiesAtQp` to be called from other compute-true materials, not by a subroutine controlled by moose.

Due to both a) and b), I don't see how we could sanely "output" properties computed by a discrete material.

**Comment by hugary1995** [comment]
One way to avoid this issue, is disallowing discrete materials to "declare" properties. But then a follow-up question is what would be the advantage of discrete material compared to the `DiscreteElementUserObkect`? Obviously we don't need duplicate systems in the framework...

**Comment by joshuahansel** [comment]
Sorry for the dumb question, but what is a "discrete material"?

**Comment by dschwen** [comment]
> Sorry for the dumb question, but what is a "discrete material"?

Not dumb. A material object that has the parameter `compute = false` and is not automatically called. These materials are usually called form within (e.g. Newton) loops within a regular material. Return mapping is a common use case. To explicitly trigger the computation of a discrete material you need to get a pointer to the Material class itself (_not_ the properties). Then you can call member functions of that class (a bit like what we do with user objects).

**Comment by dschwen** [comment]
> If MOOSE does not know how to update the properties you should write your own auxkernel to output the property, that has the custom call that you need

Then we don't need discrete Materials at all. The one thing they can do that DiscreteElementUserObjects cannot do is declare properties. The current design suggests that those properties can be output conveniently, but the system is not reliable.

**Comment by lindsayad** [comment]
Perhaps "discrete" materials as they exist now should go away, and the on-the-fly functor system should be used?

**Comment by dschwen** [comment]
Can a Functor access material properties, or take in multiple values of various types? Can a Functor compute and provide multiple values of various types?

**Comment by dschwen** [comment]
I suppose another shortcoming of DiscreteElementUserObjects is that they operate only on elements and not on faces. :-/

**Comment by GiudGiud** [comment]
Functors can do faces, in many different ways and discretizations. 

A Functor currently may not access a material property, but may retrieve as many other functors of any type you want. 
It can only compute a single value of a given type.

**Comment by dschwen** [comment]
Ok, the question is, are we using discrete materials on anything other than element interiors. For Mortar contact maybe? If so neither DiscreteElementUserObjects nor Functors are are drop in replacement. If not we could transition to DiscreteElementUserObjects fairly easily.

**Comment by lindsayad** [comment]
"FV" mortar is already using functors. "FE" mortar could just as easily

**Comment by dschwen** [comment]
> "FV" mortar is already using functors. "FE" mortar could just as easily

I don't think it would be "easy" to convert return mapping stress calculations to Functors. The question is: are those needed? @recuero?

**Comment by lindsayad** [comment]
> It can only compute a single value of a given type.

This is just as true of the traditional material system. The traditional and functor material system's primary difference is that the latter's properties are evaluated on the fly and consequently there is only ever one copy of a functor and it's functor material whereas the traditional material has three copies generally, for volume, face, and neighbor face computations

**Comment by lindsayad** [comment]
I'm not contending that the discrete material system *must* be replaced with functors. I'm just throwing out the idea. Sounds like there is another currently available option with the discrete user object

**Comment by hugary1995** [comment]
Still, we could just ban property declaration in discrete materials, and all issues will be resolved. Then we don't need to worry about face and neighbor copies.

**Comment by dschwen** [comment]
> Still, we could just ban property declaration in discrete materials, and all issues will be resolved.

Yeah, that would be a solution.

>> It can only compute a single value of a given type.

> This is just as true of the traditional material system. 

Well, if you compare Functors and Material _Properties_. But we can bundle the computation of multiple properties in a single Material.

**Comment by lindsayad** [comment]
You can define arbitrarily many functor material properties in a functor material. There should be no difference in that regard

**Comment by hugary1995** [comment]
> You can define arbitrarily many functor material properties in a functor material. There should be no difference in that regard

That's interesting, I didn't know that. Then functor material is definitely worth looking into.

**Comment by dschwen** [comment]
But each functor property would still be called individually, rather than one method that simultaneously computes multiple values. I suppose a functor could return a `std::tuple` or any data structure that would bundle multiple values, though.

**Comment by GiudGiud** [comment]
Functor material properties have caching as well, so you could leverage previously computed values. It s not necessarily going to be as efficient as computing them at the same time

**Comment by lindsayad** [comment]
> But each functor property would still be called individually, rather than one method that simultaneously computes multiple values.

What use case are you thinking about? A place where functor material properties would definitely be less efficient is in a functor material class in which a functor property is used in the definition of a lot of other functor properties. In such a situation you could be re-doing the dependency property computation quite a bit. As @GiudGiud said we do support caching, but that will up your memory requirements.

> I suppose a functor could return a `std::tuple` or any data structure that would bundle multiple values, though.

Yea we have done `std::array` and `std::vector` return types in support of @joe61vette's application

**Comment by sapitts** [comment]
Deprecating discrete materials will affect our crystal plasticity capabilities. We refactored the crystal plasticity classes away from discrete user objects to use discrete materials last year, motivated by:

- Use of many stateful material properties (the simplest model has 24 material properties; more realistic models have hundreds of stateful material properties)
- Ability to access old material properties from qps other than the current qp, for use in computing material properties at the current qp 
- Ability to combine multiple models in a modular fashion while retaining clarity in naming conventions (as in the radial return materials)

How do we move forward with the crystal plasticity capability if discrete materials are removed? Pinging @dewenyushu @amjokisaari for their input in any design discussion

**Comment by dschwen** [comment]
@sapitts thanks for bringing this up. The stateful properties declared by the discrete materials are another unique feature of the discrete material system. So we _cannot_ prohibit discrete materials from declaring properties.

I think the solution will be to 

- ensure that each discrete material is only used by _one_ regular material through `getMaterial(...)`
- add that one calling material to the list of computed materials if any of the discrete material properties is requested (e.g. for output purposes).

**Comment by hugary1995** [comment]
> - ensure that each discrete material is only used by one regular material through getMaterial(...)
> - add that one calling material to the list of computed materials if any of the discrete material properties is requested (e.g. for output purposes).

I think this will work so I have no objection.

For the second bullet point, we also need to make sure the discrete material itself is not in the list of materials to compute when outputting.

**Comment by bwspenc** [comment]
I agree with @hugary1995 -- I don't see any issues with those restrictions.

I should mention that there's another usage pattern for these 'compute=false' materials that we are working on: calling the stress calculator material from another wrapper stress calculator for plane stress. That stress calculator could in turn call another 'compute=false' material for something like a plasticity model, so we could have these models nested. It seems like that doesn't violate any of those restrictions since they're only called by a single regular material.
